File: OptionLayer.js

var OptionLayer = function () {
	this.delegate = null;
};

OptionLayer.prototype.onDidLoadFromCCB = function () {
	var winSize = cc.Director.getInstance().getWinSize();
	this.rootNode.setPosition(cc.p(winSize.width / 2 - this.rootNode.getContentSize().width / 2, winSize.height / 2 - this.rootNode.getContentSize().height / 2));
	this.rootNode.setTouchEnabled(false);
	cc.Director.getInstance().getTouchDispatcher().addTargetedDelegate(this, cc.MENU_HANDLER_PRIORITY - 1, true);
	this.updateLanguageButton();
	this.updateCCBElements();
};

OptionLayer.prototype.updateLanguageButton = function () {
	this.languageChineseMenuItem.setEnabled(true);
	this.languageEnglishMenuItem.setEnabled(true);
	this.languageJapaneseMenuItem.setEnabled(true);
	if (GameDataManager.languageType == LANGUAGE_CHINESE) this.languageChineseMenuItem.setEnabled(false);
	if (GameDataManager.languageType == LANGUAGE_ENGLISH) this.languageEnglishMenuItem.setEnabled(false);
	if (GameDataManager.languageType == LANGUAGE_JAPANESE) this.languageJapaneseMenuItem.setEnabled(false);
}

OptionLayer.prototype.updateCCBElements = function () {
	this.optionTitleLabel.setString(MultiLanguageUtil.getLocalizatedStringForKey(STRING_OPTION_TITLE));
	this.languageLabel.setString(MultiLanguageUtil.getLocalizatedStringForKey(STRING_CHOOSE_LANGUAGE));
	this.musicLabel.setString(MultiLanguageUtil.getLocalizatedStringForKey(STRING_MUSIC));
	this.soundLabel.setString(MultiLanguageUtil.getLocalizatedStringForKey(STRING_SOUND));
	this.playerNameLabel.setString(MultiLanguageUtil.getLocalizatedStringForKey(STRING_PLAYERNAME));
	this.playerName.setString(GameDataManager.playerName);
};

OptionLayer.prototype.onLanguageChinesePressed = function () {
	if (GameDataManager.languageType == LANGUAGE_CHINESE) return;
	GameDataManager.languageType = LANGUAGE_CHINESE;
	if (this.delegate) this.delegate.languageChanged();
	this.updateLanguageButton();
	this.updateCCBElements();
};

OptionLayer.prototype.onLanguageEnglishPressed = function () {
	if (GameDataManager.languageType == LANGUAGE_ENGLISH) return;
	GameDataManager.languageType = LANGUAGE_ENGLISH;
	if (this.delegate) this.delegate.languageChanged();
	this.updateLanguageButton();
	this.updateCCBElements();
};

OptionLayer.prototype.onLanguageJapanesePressed = function () {
	if (GameDataManager.languageType == LANGUAGE_JAPANESE) return;
	GameDataManager.languageType = LANGUAGE_JAPANESE;
	if (this.delegate) this.delegate.languageChanged();
	this.updateLanguageButton();
	this.updateCCBElements();
};

OptionLayer.prototype.onMusicOnPressed = function () {
	if (GameDataManager.isMusicOn) return;
	GameDataManager.isMusicOn = true;
	this.musicOnMenuItem.setEnabled(false);
	this.musicOffMenuItem.setEnabled(true);
	if (this.delegate) this.delegate.musicSettingChanged();
};

OptionLayer.prototype.onMusicOffPressed = function () {
	if (!GameDataManager.isMusicOn) return;
	GameDataManager.isMusicOn = false;
	this.musicOnMenuItem.setEnabled(true);
	this.musicOffMenuItem.setEnabled(false);
	if (this.delegate) this.delegate.musicSettingChanged();
};

OptionLayer.prototype.onSoundOnPressed = function () {
	if (GameDataManager.isSoundEffectOn) return;
	GameDataManager.isSoundEffectOn = true;
	this.soundOnMenuItem.setEnabled(false);
	this.soundOffMenuItem.setEnabled(true);
	if (this.delegate) this.delegate.soundSettingChanged();
};

OptionLayer.prototype.onSoundOffPressed = function () {
	if (!GameDataManager.isSoundEffectOn) return;
	GameDataManager.isSoundEffectOn = false;
	this.soundOnMenuItem.setEnabled(true);
	this.soundOffMenuItem.setEnabled(false);
	if (this.delegate) this.delegate.soundSettingChanged();
};

OptionLayer.prototype.onEditNamePressed = function () {
	if (GameDataManager.isSoundEffectOn) cc.AudioEngine.getInstance().playEffect("menu.mp3");
	var maxLength = 10;
	var playerName = null;
	while (!playerName || (playerName && playerName.length > maxLength)) {
		playerName = window.prompt(MultiLanguageUtil.getLocalizatedStringForKey(STRING_PROMPT_TITLE), 
								   MultiLanguageUtil.getLocalizatedStringForKey(STRING_PROMPT_INFO));
	}
	GameDataManager.updatePlayerName(playerName);
	this.playerName.setString(GameDataManager.playerName);
};

OptionLayer.prototype.onClosePressed = function () {
	if (GameDataManager.isSoundEffectOn) cc.AudioEngine.getInstance().playEffect("menu.mp3");
	cc.Director.getInstance().getTouchDispatcher().removeDelegate(this);
	this.rootNode.removeFromParent(true);
};

File: ClearFiveStages-iOS/ClearFiveStages/Scene/OptionLayer.m
#import "OptionLayer.h"
#import "GameDataManager.h"
#import "SimpleAudioEngine.h"
#import "MultiLanguageUtil.h"
#import "YIPopupTextView.h"

@implementation OptionLayer
@synthesize delegate;

-(void) didLoadFromCCB {
	CGSize winSize = [CCDirector sharedDirector].winSize;
	self.position = ccp(winSize.width / 2 - self.contentSize.width / 2, winSize.height / 2 - self.contentSize.height / 2);
	[self setTouchEnabled:YES];
	[self updateLanguageButton];
	[self updateCCBElements];
}

-(void) registerWithTouchDispatcher {
	[[CCDirector sharedDirector].touchDispatcher addTargetedDelegate:self priority:kCCMenuHandlerPriority swallowsTouches:YES];
}

-(BOOL) ccTouchBegan:(UITouch *)touch withEvent:(UIEvent *)event {
    return YES;
}

-(void) dealloc {
	if (self.delegate) [self.delegate release];
    [super dealloc];
}

// multi-language handle
-(void) updateCCBElements {
	[optionTitleLabel setString:[MultiLanguageUtil getLocalizatedStringForKey:STRING_OPTION_TITLE]];
	[languageLabel setString:[MultiLanguageUtil getLocalizatedStringForKey:STRING_CHOOSE_LANGUAGE]];
	[musicLabel setString:[MultiLanguageUtil getLocalizatedStringForKey:STRING_MUSIC]];
	[soundLabel setString:[MultiLanguageUtil getLocalizatedStringForKey:STRING_SOUND]];
	[playerNameLabel setString:[MultiLanguageUtil getLocalizatedStringForKey:STRING_PLAYERNAME]];
	[playerName setString:[GameDataManager sharedManager].playerName];
}

-(void) updateLanguageButton {
	[languageChineseMenuItem setIsEnabled:YES];
	[languageEnglishMenuItem setIsEnabled:YES];
	[languageJapaneseMenuItem setIsEnabled:YES];
	if ([GameDataManager sharedManager].languageType == LANGUAGE_CHINESE) [languageChineseMenuItem setIsEnabled:NO];
	if ([GameDataManager sharedManager].languageType == LANGUAGE_ENGLISH) [languageEnglishMenuItem setIsEnabled:NO];
	if ([GameDataManager sharedManager].languageType == LANGUAGE_JAPANESE) [languageJapaneseMenuItem setIsEnabled:NO];
}

-(void) onLanguageChinesePressed {
	if ([GameDataManager sharedManager].languageType == LANGUAGE_CHINESE) return;
	[GameDataManager sharedManager].languageType = LANGUAGE_CHINESE;
	if (self.delegate) [self.delegate languageChanged];
	[self updateLanguageButton];
	[self updateCCBElements];
}

-(void) onLanguageEnglishPressed {
	if ([GameDataManager sharedManager].languageType == LANGUAGE_ENGLISH) return;
	[GameDataManager sharedManager].languageType = LANGUAGE_ENGLISH;
	if (self.delegate) [self.delegate languageChanged];
	[self updateLanguageButton];
	[self updateCCBElements];
}

-(void) onLanguageJapanesePressed {
	if ([GameDataManager sharedManager].languageType == LANGUAGE_JAPANESE) return;
	[GameDataManager sharedManager].languageType = LANGUAGE_JAPANESE;
	if (self.delegate) [self.delegate languageChanged];
	[self updateLanguageButton];
	[self updateCCBElements];
}

-(void) onMusicOnPressed {
	if ([GameDataManager sharedManager].isMusicOn) return;
	[GameDataManager sharedManager].isMusicOn = YES;
	[musicOnMenuItem setIsEnabled:NO];
	[musicOffMenuItem setIsEnabled:YES];
    [self.delegate musicSettingChanged];
}

-(void) onMusicOffPressed {
	if (![GameDataManager sharedManager].isMusicOn) return;
    [GameDataManager sharedManager].isMusicOn = NO;
	[musicOnMenuItem setIsEnabled:YES];
	[musicOffMenuItem setIsEnabled:NO];
    [self.delegate musicSettingChanged];
}

-(void) onSoundOnPressed {
	if ([GameDataManager sharedManager].isSoundEffectOn) return;
	[GameDataManager sharedManager].isSoundEffectOn = YES;
	[soundOnMenuItem setIsEnabled:NO];
	[soundOffMenuItem setIsEnabled:YES];
    [self.delegate soundSettingChanged];
}

-(void) onSoundOffPressed {
	if (![GameDataManager sharedManager].isSoundEffectOn) return;
	[GameDataManager sharedManager].isSoundEffectOn = NO;
	[soundOnMenuItem setIsEnabled:YES];
	[soundOffMenuItem setIsEnabled:NO];
    [self.delegate soundSettingChanged];
}

-(void) onEditNamePressed {
	if ([GameDataManager sharedManager].isSoundEffectOn == YES) [[SimpleAudioEngine sharedEngine] playEffect:@"menu.mp3"];
	void(^changePlayerNameBlock)(NSString *text) = ^(NSString *text) {
											[[GameDataManager sharedManager] updatePlayerName:text];
											[playerName setString:[GameDataManager sharedManager].playerName];
	};
	YIPopupTextView* popupTextView = [[YIPopupTextView alloc] initWithPlaceHolder:@"Player name"
																		  content:@""
																		 maxCount:10
																			block:changePlayerNameBlock];	
	[popupTextView showInView:[[CCDirector sharedDirector] view]];
}

-(void) onClosePressed {
	if ([GameDataManager sharedManager].isSoundEffectOn == YES) [[SimpleAudioEngine sharedEngine] playEffect:@"menu.mp3"];
	[[CCDirector sharedDirector].touchDispatcher removeDelegate:self];
	[self removeFromParentAndCleanup:YES];
}

@end
